src/red.ts

```diff
@@ -262,157 +271,364 @@ export const serializedRedEnvSourceText = (function redEnvFactory(blueEnv: Membr
-            const { proto: blueProto } = meta;
-            // once the initialization is executed once... the rest is just noop 
-            this.initialize = noop;
-            // adjusting the proto chain of the shadowTarget (recursively)
-            const redProto = getRedValue(blueProto);
-            setPrototypeOf(shadowTarget, redProto);
-            // defining own descriptors
-            copyRedOwnDescriptors(shadowTarget, meta.descriptors);
-            // preserving the semantics of the object
-            if (meta.isFrozen) {
-                freeze(shadowTarget);
-            } else if (meta.isSealed) {
-                seal(shadowTarget);
-            } else if (!meta.isExtensible) {
-                preventExtensions(shadowTarget);
-            }
-            // future optimization: hoping that proxies with frozen handlers can be faster
-            freeze(this);
-        }
-    
-        get(shadowTarget: RedShadowTarget, key: PropertyKey, receiver: RedObject): RedValue {
-            this.initialize(shadowTarget);
-            return ReflectGet(shadowTarget, key, receiver);
-        }
-        set(shadowTarget: RedShadowTarget, key: PropertyKey, value: RedValue, receiver: RedObject): boolean {
-            this.initialize(shadowTarget);
-            return ReflectSet(shadowTarget, key, value, receiver);
-        }
-        deleteProperty(shadowTarget: RedShadowTarget, key: PropertyKey): boolean {
-            this.initialize(shadowTarget);
-            return deleteProperty(shadowTarget, key);
-        }
-        apply(shadowTarget: RedShadowTarget, redThisArg: RedValue, redArgArray: RedValue[]): RedValue {
-            const { target: blueTarget } = this;
-            this.initialize(shadowTarget);
-            let blue;
+    function getBluePartialDescriptor(redPartialDesc: PropertyDescriptor): PropertyDescriptor {
+        const bluePartialDesc = assign(create(null), redPartialDesc);
+        if ('value' in bluePartialDesc) {
+            // we are dealing with a value descriptor
+            bluePartialDesc.value = blueEnv.getBlueValue(bluePartialDesc.value);
+        }
+        if ('set' in bluePartialDesc) {
+            // we are dealing with accessors
+            bluePartialDesc.set = blueEnv.getBlueValue(bluePartialDesc.set);
+        }
+        if ('get' in bluePartialDesc) {
+            bluePartialDesc.get = blueEnv.getBlueValue(bluePartialDesc.get);
+        }
+        return bluePartialDesc;
+    }
+
+    // invoking traps
+
+    function redProxyApplyTrap(this: RedProxyHandler, shadowTarget: RedShadowTarget, redThisArg: RedValue, redArgArray: RedValue[]): RedValue {
+        const { target: blueTarget } = this;
+        let blue;
+        try {
+            const blueThisArg = blueEnv.getBlueValue(redThisArg);
+            const blueArgArray = getStaticBlueArray(redArgArray);
+            blue = blueApplyHook(blueTarget as BlueFunction, blueThisArg, blueArgArray);
+        } catch (e) {
```

Comment:
The error handling for `apply` and `construct` traps are the same. This can be extracted in a function `handleRedError(e: Error)` 

the traps become easier to read then 
```
try {
  ...
 } catch(e) {
  throw handleRedError(e);
}
```

Reply:
yeah, we did that in the next branch... we can eventually abstract that... in another PR.